### DISCUSSION TITLE
How to get populated router service in the compilerPass

### DISCUSSION THREAD

**Discussion by tacman** [symfony/symfony#54237] [Q&A] [answered]
My bundle needs to get the matching regex of existing routes that have a specific attribute  tag.  During the compiler pass, I go through the controller services, but I'm not sure how to inject the routeInterface.

```php

 // The compiler pass
public function process(ContainerBuilder $container): void
{
//   #message: "Typed property Symfony\Component\HttpKernel\Bundle\Bundle::$container must not be accessed before initialization"
//        $router = $this->container->get('router');  <--  this doesn't work

$taggedServices = $container->findTaggedServiceIds('container.service_subscriber');
foreach (array_keys($taggedServices) as $controllerClass) {
    $reflectionClass = new \ReflectionClass($controllerClass);
    foreach ($reflectionClass->getMethods() as $method) {
        foreach ($method->getAttributes(Route::class) as $attribute) {
            // @todo: get the regex that matches this route
            //  $routes = $this->router->getRouteCollection(); <-- how can I call this within the compiler pass?  Do I need us
```

I imagine I need to make sure the routes have already been loaded, there's probably some elegant way to do this but I'm not sure what it is.  Thanks!

**Comment by stof** [comment, marked as answer]
you cannot call this in a compiler pass, as compiler passes run before services are fully configured in a usable state.

You would have to delay that work to do it in a cache warmer instead.

**Reply by tacman** [reply]
Thanks!    Any recommendations on an example to follow?  Obviously, I'm starting with the documentation.

**Reply by tacman** [reply]
Maybe all of my logic that's currently in the compilerPass should be moved to warmup, since its only purpose is to write a JSON file to the cache directory for later lookups.  This was the original conversation: https://github.com/symfony/symfony/discussions/52760
